### DISCUSSION TITLE
Tests and source in the same plug-in?

### DISCUSSION THREAD

**Discussion by vogella** [eclipse-tycho/tycho#241] [General]
Having source and tests in the same artifact seems standard in Maven and Gradle projects and it would be nice to be able the same structure with Eclipse plug-ins.

I found an old discussion from 2014 about having the tests and the source in the same project here: https://www.eclipse.org/lists/tycho-user/msg05265.html Has anything changed in the meantime?  

We have a case in which we only have one project to build (and its tests). Would feel simpler to have everything in one artifact instead of setting up a multi project build.

**Comment by joeshannon** [comment]
Recently [support was added](https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#execute-unit-tests-with-eclipse-plugin-packaging) for running tests as part of an eclipse-plugin artifact.

It looks as though this supports non-plugin tests with the maven-surefire-plugin in addition to plugin tests (at the integration-test phase) but I haven't tried it personally yet.

**Reply by laeubi** [reply]
> It looks as though this supports non-plugin tests with the maven-surefire-plugin in addition to plugin tests (at the integration-test phase)

Correct, see https://github.com/eclipse/tycho/tree/master/tycho-its/projects/surefire.combinedtests/bundle.test for an example.

**Comment by vogella** [comment]
Thanks @joeshannon and @laeubi 

Trying to build tycho-its results in an error for me. Opened #246 for that.

**Reply by laeubi** [reply]
You can also run this specific integrationtest by cd into `tycho-its/projects/surefire.combinedtests/bundle.test ` and then run `mvn clean install -Dtycho-version=2.4.0` with the current release.

**Comment by kursat-ozer** [comment]
Hi,
I read this feature introduced in version 2.4.0.
We are using java8 in our project and we cant migrate to java11 so we cant use tyhco 2.x.x. versions. Is there any way to add this feaure to tycho versions 1.x.x ?

**Reply by laeubi** [reply]
You just need java 11 to *run* tycho the code you can still use java-8 and can even be compiled using a jdk8 jvm.

**Reply by kursat-ozer** [reply]
I didnt understand what you mean. Can you please elaborate more?

**Reply by Bananeweizen** [reply]
run maven from a Java 11 path. have source=8, target=8 in your pom for all relevant compilation targets. that way tycho runs on 11, but your build still uses 8.
it might also be necessary to configure maven toolchains.xml, not sure of that.
